### PR TITLE
Python: add BLPOP and BRPOP commands

### PR DESCRIPTION
**Pull request by aaron-congo**
*Issue #, if available:*
N/A

*Description of changes:*
- Add `BLPOP` command: https://redis.io/commands/blpop/
- Add `BRPOP` command: https://redis.io/commands/brpop/

Note: depends on https://github.com/aws/glide-for-redis/pull/1239

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/python/glide/async_commands/core.py [hunk 1]
```diff
@@ -765,6 +765,37 @@ async def lpop_count(self, key: str, count: int) -> Optional[List[str]]:
...
+
+        See https://redis.io/commands/blpop/ for details.
+
+        Note: BLPOP is a client blocking command, see
+        https://github.com/aws/glide-for-redis/wiki/General-Concepts#blocking-commands for more details and best
+        practices.
+
+        Args:
+            keys (List[str]): The keys of the lists to pop from.
+            timeout (float): The number of seconds to wait for a blocking BLPOP operation to complete. A
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:782]:
```suggestion
            timeout (float): The number of seconds to wait for a blocking operation to complete. A
```
San proposed to remove command name there - I'll do for all blocking commands I did in java

File: python/python/glide/async_commands/transaction.py [hunk 2]
```diff
@@ -618,6 +618,30 @@ def lpop_count(self: TTransaction, key: str, count: int) -> TTransaction:
...
+
+        See https://redis.io/commands/blpop/ for details.
+
+        Note: BLPOP is a client blocking command, see
+        https://github.com/aws/glide-for-redis/wiki/General-Concepts#blocking-commands for more details and best
+        practices.
+
+        Args:
+            keys (List[str]): The keys of the lists to pop from.
+            timeout (float): The number of seconds to wait for a blocking BLPOP operation to complete. A
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/transaction.py:635]: same

File: python/python/glide/async_commands/core.py [hunk 3]
```diff
@@ -765,6 +765,37 @@ async def lpop_count(self, key: str, count: int) -> Optional[List[str]]:
...
+                value of `0` will block indefinitely.
+
+        Returns:
+            Optional[List[str]]: A list containing the `key` from which the element was popped and the `value` of the
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:786]: Should we add `A two-element list...`?

File: python/python/tests/test_async_client.py [hunk 4]
```diff
@@ -909,6 +909,44 @@ async def test_rpushx(self, redis_client: TRedisClient):
         with pytest.raises(RequestError):
             await redis_client.rpushx(key2, [])
 
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_blpop(self, redis_client: TRedisClient):
+        key1 = f"{{test}}-f{get_random_string(10)}"
+        key2 = f"{{test}}-f{get_random_string(10)}"
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:916]: nit: this simplifies debugging
```suggestion
        key1 = f"{{test}}-1-f{get_random_string(10)}"
        key2 = f"{{test}}-2-f{get_random_string(10)}"
```

File: python/python/tests/test_async_client.py [hunk 5]
```diff
@@ -909,6 +909,44 @@ async def test_rpushx(self, redis_client: TRedisClient):
...
+        value1 = "value1"
+        value2 = "value2"
+        value_list = [value1, value2]
+
+        assert await redis_client.lpush(key1, value_list) == 2
+        assert await redis_client.blpop([key1, key2], 0.5) == [key1, value2]
+
+        # Redis versions < 7.0.0 do not accept floating point values for the timeout, so will be rounded
+        timeout = 1.0 if check_if_server_version_lt(redis_client, "7.0.0") else 0.001
+        assert await redis_client.set("foo", "bar")
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:926]: You missed the test
```suggestion
        assert await redis_client.blpop([key2], timeout) == Null

        assert await redis_client.set("foo", "bar")
```

File: python/python/tests/test_async_client.py [hunk 6]
```diff
@@ -1868,7 +1906,14 @@ class TestExceptions:
     async def test_timeout_exception_with_blpop(self, redis_client: TRedisClient):
         key = get_random_string(10)
         with pytest.raises(TimeoutError):
-            await redis_client.custom_command(["BLPOP", key, "1"])
+            await redis_client.blpop([key], 1)
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_timeout_exception_with_brpop(self, redis_client: TRedisClient):
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:1913]: you can embed this test into `test_brpop` as I proposed for blpop

**Comment by Yury-Fridlyand** [comment]
Pick doc changes from #198

**Review by Yury-Fridlyand — APPROVED** [review]

File: python/python/tests/test_async_client.py [hunk 7]
```diff
@@ -909,6 +909,46 @@ async def test_rpushx(self, redis_client: TRedisClient):
...
+        value1 = "value1"
+        value2 = "value2"
+        value_list = [value1, value2]
+
+        assert await redis_client.lpush(key1, value_list) == 2
+        assert await redis_client.brpop([key1, key2], 0.5) == [key1, value1]
+
+        # Redis versions < 7.0.0 do not accept floating point values for the timeout, so will be rounded
+        timeout = 1.0 if check_if_server_version_lt(redis_client, "7.0.0") else 0.001
+        assert await redis_client.brpop([key2], timeout) is None
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:946]: don't run this test on redis 6

**Review by jonathanl-bq — APPROVED** [review]

**Comment by jonathanl-bq** [comment]
Approved on condition that you make Yury's suggested changes around Redis versioning.

File: CHANGELOG.md [hunk 8]
```diff
@@ -15,6 +15,7 @@
 * Python: Added ZREMRANGEBYLEX command ([#1306](https://github.com/aws/glide-for-redis/pull/1306))
 * Python: Added LINSERT command ([#1304](https://github.com/aws/glide-for-redis/pull/1304))
 * Python: Added GEOPOS command ([#1301](https://github.com/aws/glide-for-redis/pull/1301))
+* Python: Added BLPOP and BRPOP commands (TODO: add PR link)
```
Comment (Yury-Fridlyand) [inline CHANGELOG.md:18]: todo

File: python/python/tests/test_async_client.py [hunk 9]
```diff
@@ -2233,6 +2279,23 @@ async def test_cluster_fail_routing_by_address_if_no_port_is_provided(
             await redis_client.info(route=ByAddressRoute("foo"))
 
 
+@pytest.mark.asyncio
+class TestExceptions:
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_timeout_exception_with_blpop(self, redis_client: TRedisClient):
+        key = get_random_string(10)
+        with pytest.raises(TimeoutError):
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:2288]: should not raise an error anymore with any timeout AFAIK
I want to add such test for all blocking commands

File: python/python/tests/test_async_client.py [hunk 10]
```diff
@@ -853,6 +853,31 @@ async def test_lpushx(self, redis_client: TRedisClient):
...
+        key1 = f"{{test}}-1-f{get_random_string(10)}"
+        key2 = f"{{test}}-2-f{get_random_string(10)}"
+        value1 = "value1"
+        value2 = "value2"
+        value_list = [value1, value2]
+
+        assert await redis_client.lpush(key1, value_list) == 2
+        assert await redis_client.blpop([key1, key2], 0.5) == [key1, value2]
+
+        assert await redis_client.blpop(["non_existent_key"], 0.5) is None
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:868]: add a comment - .5 is higher than client request timeout
add a test with 0 timeout, but wrap it with something like `await asyncio.wait_for(future, timeout=5)` to avoid real blocking

File: python/python/tests/test_async_client.py [hunk 11]
```diff
@@ -911,6 +936,31 @@ async def test_rpushx(self, redis_client: TRedisClient):
...
+        key1 = f"{{test}}-1-f{get_random_string(10)}"
+        key2 = f"{{test}}-2-f{get_random_string(10)}"
+        value1 = "value1"
+        value2 = "value2"
+        value_list = [value1, value2]
+
+        assert await redis_client.lpush(key1, value_list) == 2
+        assert await redis_client.brpop([key1, key2], 0.5) == [key1, value1]
+
+        assert await redis_client.brpop(["non_existent_key"], 0.5) is None
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:951]: same here

File: python/python/tests/test_async_client.py (hunk 10 above)
Comment (Yury-Fridlyand) [reply python/python/tests/test_async_client.py:868]: https://github.com/Bit-Quill/glide-for-redis/pull/209/files#diff-5db9a3bd41d647cfdba521ea248ea07d7c78bc9aaa41a26bb923d4f92a3efdd2R1691-R1699

File: .github/workflows/pypi-cd.yml
Comment (Yury-Fridlyand) [inline .github/workflows/pypi-cd.yml:1]: these 2 could be a part of another PR

**Review by Yury-Fridlyand — APPROVED** [review]
GTG to upstream without CI part
This should be fixed in another PR and merged ahead of code changes.

**Review by Yury-Fridlyand — APPROVED** [review]
